### PR TITLE
Progress reporting for #r nuget

### PR DESCRIPTION
**Pull request by rchande**

### CODE REVIEW

File: WorkspaceServer.Tests/Kernel/CSharpKernelTests.cs [hunk 1]
```diff
@@ -410,6 +410,16 @@ public async Task When_SubmitCode_command_adds_packages_to_csharp_kernel_then_Pa
                                .ToEnumerable()
                                .ToArray();
 
+            events
+                .First()
+                .Should()
+                .Match(e => e is DisplayedValueProduced && ((DisplayedValueProduced)e).Value.ToString().Contains("Attempting to install"));
```
Comment (colombod) [inline WorkspaceServer.Tests/Kernel/CSharpKernelTests.cs]: probably here we should assert on a sequence of events

File: WorkspaceServer/Kernel/CSharpKernelExtensions.cs [hunk 2]
```diff
@@ -68,22 +69,50 @@ public static CSharpKernel UseNugetDirective(this CSharpKernel kernel, INativeAs
                 {
                     Handler = async context =>
                     {
-                        var refs = await restoreContext.AddPackage(package.PackageName, package.PackageVersion);
+                        var message = $"Attempting to install package {package.PackageName}";
+                        if (!string.IsNullOrWhiteSpace(package.PackageVersion))
+                        {
+                            message += $", version {package.PackageVersion}";
+                        }
```
Comment (jonsequitur) [inline WorkspaceServer/Kernel/CSharpKernelExtensions.cs]: Can we also get a notification about which actual version was installed in the case where the version was not specified?

File: WorkspaceServer/Kernel/CSharpKernelExtensions.cs [hunk 3]
```diff
@@ -68,22 +69,50 @@ public static CSharpKernel UseNugetDirective(this CSharpKernel kernel, INativeAs
...
 
-                            kernel.AddMetadataReferences(refs);
+                            kernel.AddMetadataReferences(result.References);
+
+                            context.Publish(new DisplayedValueProduced($"Successfully added reference to package {package.PackageName}", context.Command));
+                            context.Publish(new NuGetPackageAdded(package));
+                        }
+                        else
+                        {
+                            context.Publish(new DisplayedValueProduced($"Failed to add reference to package {package.PackageName}", context.Command));
```
Comment (jonsequitur) [inline WorkspaceServer/Kernel/CSharpKernelExtensions.cs]: When it fails, why does it fail, and is there additional information we can surface?

File: WorkspaceServer/Packaging/PackageRestoreContext.cs [hunk 4]
```diff
@@ -4,12 +4,25 @@
...
 using Clockwise;
 using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.VisualBasic.Syntax;
 using MLS.Agent.Tools;
 
 namespace WorkspaceServer.Packaging
 {
     public class PackageRestoreContext
     {
+        public class AddReferenceResult
```
Comment (jonsequitur) [inline WorkspaceServer/Packaging/PackageRestoreContext.cs]: Any reason for this to be a nested class?

File: WorkspaceServer/Kernel/CSharpKernelExtensions.cs [hunk 5]
```diff
@@ -63,23 +64,52 @@ public static CSharpKernel UseNugetDirective(this CSharpKernel kernel, INativeAs
...
                 addPackage.Handler = async context =>
                 {
-                    var refs = await restoreContext.AddPackage(package.PackageName, package.PackageVersion);
+                    var message = $"Attempting to install package {package.PackageName}";
+                    if (!string.IsNullOrWhiteSpace(package.PackageVersion))
+                    {
+                        message += $", version {package.PackageVersion}";
+                    }
+
+                    var key = message;
```
Comment (jonsequitur) [inline WorkspaceServer/Kernel/CSharpKernelExtensions.cs]: Is the redundant variable needed?

Comment (jonsequitur) [inline WorkspaceServer/Kernel/CSharpKernelExtensions.cs]: Maybe call this `valueId` so it's clearer what its significance is.

File: WorkspaceServer/Kernel/CSharpKernelExtensions.cs (hunk 2 above)
Comment (rchande) [reply WorkspaceServer/Kernel/CSharpKernelExtensions.cs]: Will do this in a subsequent change

**Pull request merged**